Fix MergeRequestBuildActionTest

pdudits opened the pull request:
Call to MergeRequestBuildAction.execute always thrown an exception and so mock
verifications were never performed - the tests passed because of ExpectedException rule.

All the decisions on whether or not to build a project are inside mergeRequestHookTriggerHandler,
therefore mocking GitLabPushTrigger didn't exercise any of the build triggering logic, and so some
of the tests now fail, demonstrating actual issues.

Fixes #734.

Commit:
Fix MergeRequestBuildActionTest

Call to MergeRequestBuildAction.execute always thrown an exception and so mock
verifications were never performed - the tests passed because of ExpectedException rule.

All the decisions on whether or not to build a project are inside mergeRequestHookTriggerHandler,
therefore mocking GitLabPushTrigger didn't exercise any of the build triggering logic, and so some
of the tests now fail, demonstrating actual issues.

pdudits commented:
I will eventually  extend the pull request to fix #734, for now I'd like to know if there's cleaner way to check whether build would run, instead of using `Mockito.spy`

omehegan commented:
@Argelbargel any chance you can answer @pdudits's question here?

omehegan commented:
@pdudits it's interesting that you discovered some of the tests are invalid. I was poking around at #705 yesterday. I can reproduce the issue. I found that we have tests at e.g. https://github.com/jenkinsci/gitlab-plugin/blob/master/src/test/java/com/dabsquared/gitlabjenkins/webhook/build/PushBuildActionTest.java which pass, so I couldn't figure out why that would be if the feature is broken. But I am not too confident that I am understanding the test code correctly when I read it.

Argelbargel commented:
@pdudits: the only way i know is to use the [Jenkins Test Harness](https://wiki.jenkins.io/display/JENKINS/Unit+Test). But using that means that you have to programmatically create all the required structures (projects and such) and executing the tests is rather slow. I think this is the reason why mockito is used so much for testing.

What you would do in this case is triggering the build and then checking that the queue (`jenkins.getInstance().getQueue()`) contains an appropriate run. Still a bit tricky as you've got no real control about how quickly the build is stepping through the queue's lifecyle and it will be processed concurrently to your test-code. So you'll have to implement an register a QueueListener and wait for it to be triggered. Which means littering the code with Thread.sleep(). Technically the better way to write such tests in my opinion, but does not generate more readable test-code :-/

omehegan commented:
@Argelbargel @pdudits I asked one of the longtime contributors to Jenkins how he would handle this case, and he agrees that using the JTH is probably the "best" way. 

"Yeah, best answer I've got is to use JenkinsRule and not use mocks - and then maybe have something hit the webhook endpoint on the Jenkins instance with what's expected from gitlab?" Hope we can come up with something that works.

pdudits commented:
Thanks. As far as I understand the code, we are already launching Jenkins and creating the project programmatically, so I'll have a look on QueueListener.

omehegan commented:
@pdudits do you think you will be able to work on this anytime soon? Just trying to organize PRs for a release.

pdudits commented:
Oh, the time flies quickly! :)

I'm afraid this will need to skip next release, because even when I clean that up, it still exposes some bugs and therefore makes tests fail.

I'll try to have this cleaned up thoughout June.

omehegan commented:
@pdudits thanks for your help, I really appreciate it!

omehegan commented:
@pdudits still hoping you can finish this in the near future, any chance?

pdudits commented:
@omehegan Unfortunately it needs to remain open as reminder of fluctuating spare-time-coding interests :). I go offline for two weeks now, so I again only can promise "maybe end of next month".  

But if someone has an idea about that [failing test](https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fgitlab-plugin/detail/PR-742/5/tests), especially about its original intent, it would be great if they can contribute.

omehegan commented:
@pdudits ping :)

Commit:
Use QueueListener instead of Mockito.spy

Commit:
Use chain of general OR conditions for deciding wether merge request hook would fire.

Based on the issues reported it is more manageable to handle conditions for each checkbox
separately.

Commit:
Add test and fix for #732

pdudits commented:
All right, all right :). I do have some work in progress stuff, so I'll at least rebase it and run it against CI here.

Commit:
Raise timeout for test

It could take more time on some platforms to startup Jenkins.

pdudits commented:
Concerning failing test
`MergeRequestBuildActionTest.skip_alreadyBuiltMR`  it appears, that this feature was effectively removed with c328784c79adf00d9588b90135f7d6a589bdd707 in 1.5.7, as open requests are always built.

omehegan commented:
@pdudits it's interesting that the test only fails on Windows, and did not fail before when we removed the feature! Whaaa? But you are right that it was removed, so it would seem like we do not need the test.

pdudits committed:
Make test fail consistently on Windows and Linux by raising the timeout of JenkinsRule

pdudits committed:
Clean up MergeRequestHookTriggerHandlerImplTest

Made the legacy constructor have the same behaviour as old code to fix most of the tests.
Gradually, all of the test cases should use the factory rather than using the constructor, as that is the code that is actually used by the plugin

pdudits committed:
Fix the case "approved merge request should be built if and only if triggerOnApprovedMergeRequest is set"

TriggerConfigChain now also has exclusion rules - if flag is not set, and the trigger condition is satisfied, then the trigger will not fire.
This allows to crate condition "if and only if" for approved merge requests.

pdudits committed:
Use trigger configuration for tests of MergeRequestHookTriggerHandlerImpl

The parameters have been extracted to a common interface, so they're approachable more easily than 7 distinct parameters.
Some of the test cases had to be adapted, because actual trigger checks for actions for closing and accepting rather than on state.

pdudits committed:
Fix testcases for already built commits.

The feature works, just needs to be triggered appropriately.

pdudits commented:
So. I think this may be the ultimate merge request trigger fix 😁

omehegan commented:
@pdudits fantastic! Thank you for sticking with this, I really appreciate the contribution! Am I understanding correctly that this PR now fixes #734 as well as the associated tests?

pdudits commented:
Yes, this fixes #734, as proved by the testcases in `MergeRequestHookTriggerHandlerImplTest`. These now directly test checkboxes' combinations against webhook payloads.

omehegan commented:
This was released today in 1.5.10.
